### ISSUE TITLE
used_by has no effect

### ISSUE THREAD

**Issue by colinxs** [nvim-treesitter/nvim-treesitter#1317]
**Describe the bug**

Setting `parser_config.lang.used_by = { ... }` has no effect.

**To Reproduce**

With the bash parser installed and checkhealth all green, the following init.vim is sufficient:
```
local parser_config = require "nvim-treesitter.parsers".get_parser_configs()

-- Bash parser doesn't work well for ZSH
parser_config.bash.used_by = { "PKGBUILD" }
require'nvim-treesitter.configs'.setup {
  ignore_installed = { "bash" },
  highlight = {
    enable = true, 
  },  
}
parser_config.bash.used_by = { "PKGBUILD" }
EOF
```
Now go open a .zsh file. It will be highlighted via treesitter. I tried setting it before and after but the result is the same: zsh files are highlighted with treesitter.

**Expected behavior**

For treesitter to be disabled on zsh files.

**Output of `:checkhealth nvim_treesitter`**

<details>
<code>
========================================================================
## Installation
  - OK: tree-sitter found  0.19.3 (parser generator, only needed for :TSInstallFromGrammar)
  - OK: node found v14.16.1 (only needed for :TSInstallFromGrammar)
  - OK: git executable found.
  - OK: /nix/store/zzvq5qwlm2xikawfqxb0q8gl2bw391a9-gcc-wrapper-10.2.0/bin/gcc executable found. Selected from { "/nix/store/zzvq5qwlm2xikawfqxb0q8gl2bw391a9-gcc-wrapper-10.2.0/bin/gcc", "cc", "gcc", "clang", "cl" }
  - OK: Neovim was compiled with tree-sitter runtime ABI version 13 (required >=13). Parsers must be compatible with runtime ABI.

## Parser/Features H L F I
  - toml           ✓ ✓ ✓ ✓
  - go             ✓ ✓ ✓ ✓
  - bibtex         ✓ . ✓ ✓
  - bash           ✓ ✓ ✓ .
  - latex          ✓ . ✓ .
  - ruby           ✓ ✓ ✓ ✓
  - jsonc          ✓ ✓ ✓ ✓
  - r              ✓ ✓ . .
  - elm            . . . .
  - php            x x ✓ ✓
  - java           ✓ ✓ . ✓
  - c              ✓ ✓ ✓ ✓
  - javascript     ✓ ✓ ✓ ✓
  - html           ✓ ✓ ✓ ✓
  - json           ✓ ✓ ✓ ✓
  - regex          ✓ . . .
  - css            ✓ . ✓ ✓
  - comment        ✓ . . .
  - fennel         ✓ ✓ . .
  - yaml           ✓ ✓ ✓ ✓
  - ocaml          ✓ ✓ ✓ .
  - cpp            ✓ ✓ ✓ ✓
  - python         ✓ ✓ ✓ ✓
  - lua            ✓ ✓ ✓ ✓
  - rust           ✓ ✓ ✓ ✓
  - c_sharp        ✓ . . .
  - typescript     ✓ ✓ ✓ ✓
</code>
</details>

**Output of `nvim --version`**

<details>
<code>

NVIM v0.5.0-dev
Build type: Release
LuaJIT 2.1.0-beta3
Compilation:
Compiled by nixbld

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "
/nix/store/6h5ihva8d6icb6chj6gib6a2d3h15wqq-neovim-unwrapped-master/share/nvim
"

</code>
</details>

**Additional context**

**Comment by colinxs** [comment]
Update: if I had `packadd nvim-treesitter` right above the above block then everything works fine. Not sure why that is? I get no errors without the `packadd` and the parsers work just fine, it's only the config which doesn't persist. Odd.

**Comment by colinxs** [comment]
So if I add `packadd nvim-treesitter` right above that lua block then zsh is no longer highlighted with treesitter. Not sure why that is, as the rest of the config and parsers work just fine with out it, it's just setting `used_by` that's broken.

**Comment by colinxs** [comment]
Spoke too soon, that doesn't fix the issue.

**Comment by vigoux** [comment]
The `used_by` key in parser config is actually referring to filetypes that should use this parser.
Are you sure that, when opening a `PKGBUILD` file, the filetype is `PKGBUILD`?

**Comment by colinxs** [comment]
Not sure. It's only there because [it's the default](https://github.com/nvim-treesitter/nvim-treesitter/blob/a01edfbdfbda6e68d4d003cdb9735f2a7f4bd4ea/lua/nvim-treesitter/parsers.lua#L152). My only goal is to disable the bash parser for zsh because it's broken.

**Comment by vigoux** [comment]
BTW, your wrote `ignore_installed` in your example config, but it should be `ensure_installed`

**Comment by colinxs** [comment]
I've installed the parsers via other means (with the exact same for rev's, so that's not the cause of the issue), so I think ignore_installed is correct in my case. I don't want nvim-treesitter to install anything but without ignore_installed it won't load them.

**Comment by theHamsta** [comment]
@colinxs is your `PKGBUILD` really of filetype `PKGBUILD`? For me `PKGBUILD` files have the filetype `sh` (`:echo &ft`). It's also possible that you are affected by https://github.com/nvim-treesitter/nvim-treesitter/issues/1249 and the `PKGBUILD` filetype gets set to late.

**Comment by colinxs** [comment]
This isn't about PKGBUILD as I mentioned [above] (https://github.com/nvim-treesitter/nvim-treesitter/issues/1317#issuecomment-843752568). I just want to disassociate the bash parser from zsh, so I took the default `used_by` linked above and subtracted zsh, leaving PKGBUILD.

**Comment by theHamsta** [comment]
Subtracting will not work since some weird logic is happening that is registering the parser. Maybe this would work

`list.PKGBUILD = list.bash`

**Comment by carlitux** [comment]
@colinxs `used_by` to update configs is intended for simple use cases and allows only one file type per sentence.

must be

```lua
local parser_config = require "nvim-treesitter.parsers".get_parser_configs()

parser_config.bash.used_by =  "PKGBUILD" 

require'nvim-treesitter.configs'.setup {
  ignore_installed = { "bash" },
  highlight = {
    enable = true, 
  },  
}
```

hope this helps

**Comment by colinxs** [comment]
Again, I don't care about PKGBUILD highlighting at all. I just don't want the bash parser being used for zsh because it works poorly. Said another way, I want to change the default from:
```
list.bash = {
  install_info = {
    url = "https://github.com/tree-sitter/tree-sitter-bash",
    files = { "src/parser.c", "src/scanner.cc" },
  },
  used_by = { "zsh", "PKGBUILD" },
  filetype = 'sh',
  maintainers = {"@TravonteD"},
}
```
To:
```
list.bash = {
  install_info = {
    url = "https://github.com/tree-sitter/tree-sitter-bash",
    files = { "src/parser.c", "src/scanner.cc" },
  },
  used_by = {"PKGBUILD" }, # Only difference is this line
  filetype = 'sh',
  maintainers = {"@TravonteD"},
}
```
"PKGBUILD" is only their because it was part of the default. You could have said `used_by` to `{}` for all I care.

Any combination of the commented lines below, whether it be 1, 2, or all of the lines, in any order,  appear to have zero effect:
```
# parser_config.bash = { filetype = 'sh', used_by = { "PKGBUILD" } }
# parser_config.bash.used_by = "PKGBUILD"
require'nvim-treesitter.configs'.setup { ... }
# parser_config.bash = { filetype = 'sh', used_by = { "PKGBUILD" } }
# parser_config.bash.used_by = "PKGBUILD"
```

In every single case, the default `used_by = { "zsh", "PKGBUILD" }` is in effect and the bash parser used for zsh files.

So I'm not really sure what's going on.

**Comment by colinxs** [comment]
So the issue is `parsers.ft_to_lang` never gets updated.

`get_parser_configs` correctly shows:
```
bash = {
    filetype = "sh",
    used_by = {},
  },
```
after 
```
parser_config.bash = { filetype = 'sh', used_by = {} }
```
while `get_buf_lang` still reports "bash" on a zsh file (and set ft? correctly reports zsh).

I'm new to lua but judging from the logic [here](https://github.com/nvim-treesitter/nvim-treesitter/blob/0efa1ec83ad8a6990e35db184cfbb3f6818e6e94/lua/nvim-treesitter/parsers.lua#L15) your __newindex method/function/foobar doesn't remove the old mapping.

You would need to:
1) Get the current used_by value
2) Set difference that with the new used_by value
3) Delete all the mapping in the old used_by that aren't in the new used_by.

You need to do the same for filetype as well.

**Comment by carlitux** [comment]
@colinxs you just need to avoid using `bash` config and add another based on bash config for `PKGBUILD` or whatever you want to name it.

The `filetype` and `used_by` are what you need to create your own config. There is no way to remove from `parsers.ft_to_lang`.

Here the doc

https://github.com/nvim-treesitter/nvim-treesitter#adding-parsers

also take care that you typed `ignore_installed = { "bash" }` and docs says `ignore_install = { "bash" },`

**Comment by colinxs** [comment]
Okay, forget about PKGBUILD altogether. As I said earlier it only came up because it was the verbatim default, but the point is that the bash parser is used for zsh, and I want to disable that. 

Creating a new config does absolutely nothing to address that. I don't want `zsh`in `ft_to_lang` in the first place, but it's there because of the default parser config for `zsh`. The `bash` parser is completely borked for even mildly complicated `zsh` code. Since I cannot directly delete from `ft_to_parsername` because of its local scope the only thing I could conceivably do is overwrite it. But with what? There's no ZSH parser.

It sounds like an anti-feature if I can't override a default configuration. 

It's certainly a bug that `update_ft_to_parsername`  doesn't delete the old association. You mean to say the following is a good interface?
```
parser_config.bash = nil
require'nvim-treesitter.configs'.setup { ... }
# bash parser still operational
```

**Comment by carlitux** [comment]
After reading carefully all the messages into the thread the goal is to disable bash, not to modify config. That is why all answers we give you are wrong.

your issue must be your setup

```lua
require'nvim-treesitter.configs'.setup {
  ignore_install = { "bash" }, -- you have ignore_installed
  highlight = {
    enable = true, 
  },  
}
```

try that, please

I tested with this config, setting bash into ensure_installed but also setting bash into ignore_install

```lua
treesitter_config.setup {
  ignore_install = {'bash'},
  highlight = {
    enable = true,
     use_languagetree = true
  },
  ensure_installed = {
    "bash",
    "comment",
    "css",
    "graphql",
    "html",
    "java",
    "javascript",
    "jsdoc",
    "json",
    "jsonc",
    "lua",
    "php",
    "python",
    "regex",
    "scss",
    "toml",
    "tsx",
    "typescript",
  }
}
```

and running checkhealth bash is not installed

![Screenshot from 2021-05-19 21-34-38](https://user-images.githubusercontent.com/40538/118905778-6584eb80-b8ea-11eb-9919-320740c4bdd0.png)

Maybe you need to uninstall later bash not sure about this.

Hope this helps you

**Comment by colinxs** [comment]
Nope. As I said in the very first post and every single post since:

> **Expected behavior**
> 
> For treesitter to be disabled on zsh files.

Put differently, I do not want the bash *parser* used for the zsh *filetype*, but I still want the bash *parser* used for the *bash* filetype.

**Comment by carlitux** [comment]
If you want to deal with the issue till the interface is changed, to be able to change default configs. I don't know if that will happens, I requested a long time ago similar to this https://github.com/nvim-treesitter/nvim-treesitter/issues/31.

If you want to get solved you have to:

1. Disable default bash
2. Create your own bash config

```lua
local parser_config = require "nvim-treesitter.parsers".get_parser_configs()

-- custom bash that will not be used by zsh
parser_config.mybash = {  
  install_info = {    
    url = "https://github.com/tree-sitter/tree-sitter-bash",    
    files = { "src/parser.c", "src/scanner.cc" },
  },  
  used_by = {},  
  filetype = 'sh',  
}

require'nvim-treesitter.configs'.setup {
  ignore_install = { "bash" },  -- disable bash
  highlight = {
    enable = true, 
  },  
}
```

Here my tests getting this bash TS enable, zsh with syntax

![Screenshot from 2021-05-19 21-55-39](https://user-images.githubusercontent.com/40538/118907194-1ab8a300-b8ed-11eb-8cea-5426febd63ca.png)

**Comment by carlitux** [comment]
Would be good a PR to reindex `ft_to_parsername` when modifying `used_by` instead of adding new filetypes

**Comment by colinxs** [comment]
I really think nvim-treesitter should follow the API design of lsp/nvim-lspconfig:
- Don't manage the installation of parsers. Move that code elsewhere and point to it in the README for users, but don't make the assumption that I want you installing anything. That let's you focus on the important aspects of treesitter <-> neovim integration (queries, highlight groups, plugins, etc) and not on being a package manager (relevant: https://medium.com/@sdboyer/so-you-want-to-write-a-package-manager-4ae9c17d9527).
- Provide [default configurations](https://github.com/neovim/nvim-lspconfig/blob/d23816b7b00028375e53dba2aa30f2b56573cf54/lua/lspconfig/rnix.lua#L8) that **only** get used if the users asks for a specific parser to be loaded **and** did not provide any config themselves. 

So something like:
```
require'nvim-treesitter'.setup({
    default_config = {
        parsers = {
            path = "~/.local/share/nvim/site/parsers"
            highlight = true
            indent = false;
        }
    }
    parsers = {
        bash = {
            -- default path
            highlight = false -- override default
            -- default indent
            -- didn't set "used_by", so fallback to nvim-treesitter.defaults.bash.used
        }
    }
})
```

But hey I'm just some random guy on the internet.

**Comment by colinxs** [comment]
@carlitux An unfortunate side-effect of that hack is it breaks TSModuleInfo/checkhealth/etc... :/

**Comment by colinxs** [comment]
Until a proper fix you might consider adding this so folks can disable filetype<-->parser associations.

```
From bef0379f22bb2a83b1c03c5946df3e9c09eacd33 Mon Sep 17 00:00:00 2001
From: Colin Summers <me@colinxsummers.com>
Date: Wed, 19 May 2021 23:00:05 -0700
Subject: [PATCH] allow overriding ft_to_parsername

---
 lua/nvim-treesitter/parsers.lua | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/lua/nvim-treesitter/parsers.lua b/lua/nvim-treesitter/parsers.lua
index 95c5287..676f58a 100644
--- a/lua/nvim-treesitter/parsers.lua
+++ b/lua/nvim-treesitter/parsers.lua
@@ -593,7 +593,8 @@ list.fortran = {
 }
 
 local M = {
-  list = list
+  list = list,
+  ft_to_parsername = ft_to_parsername,
 }
 
 function M.ft_to_lang(ft)
-- 
2.31.1

```

**Comment by theHamsta** [comment]
> Don't manage the installation of parsers. 

This is completely to unrelated to the issue. I don't get why you would want to break a working solution that is testable by a readme that is not testable and probably not working. We can't provide versioned scm files when we don't manage the parser version. If you don't like the feature the just don't use it. We've put effort to accept parsers no matter how they are installed (use either npm, Tree-Sitter CLI or invoke the compiler yourself, nvim has documentation on where to put the compiler binary). We don't want everyone to manage all parsers versions by themselves. I don't think we have any issues with parser installation right now or any maintenance burden on our side.

There is a simple solution to the problem: just remove used_by all together or at least to don't expose it to the user and remove all the metatable code. Users should just be able to set a ft_to_lang in the setup call.

**Comment by theHamsta** [comment]
There will probably also be a grand unified solution to install parsers with Tree-Sitter 1.0. They will be probably downloadable packages that can be hosted via github releases.

**Comment by vigoux** [comment]
So, first of all, there's no need to get angry, both on @theHamsta side and @colinxs.

Second, please, contributors and maintainers, please read the issues _correctly_ and stop assuming people are dumb and don't know how to use the plugin.

Third, this is a _bug_ originating in a  _wrong design choice_ that we are suffering from, since we started to try to install parsers. Though, this is work that we carried a crap ton of work on this, since more than a year (and @theHamsta has been working on it a _lot_), so you can't just say "plz don't do it" without considering the amount of work behind it.

Now to try to actually fix the issue, I think we need a complete rework of the configuration / installation process.

The installation can be managed by some special module for example, and included in our "standard" distribution of plugin. The thing is, installing a parser is much much much harder than installing a language server in most cases, and requires a quite good understanding on how tree-sitter actually works internally.

For the configuration part, we know there is problems with it, and we actually discussed about reworking it. I think that this issue is an alarm, that should warm us that, with neovim 0.5 approaching dangerously, we should maybe break some things, to build a sane, and easy, to grasp, in order to simplify maintenance in the future (and I like some ideas from @colinxs ).

TLDR: we need to rework the whole setup / installation code, as it is too complex.

**Comment by colinxs** [comment]
Thanks for the comments everyone, really. It's refreshing to get such a hearty and timely response. Apologies if I was rude in any way. I think we all want the same thing here (and are perhaps a bit eager for the long-awaited 0.5 release :) ).

On the installation question, don't get me wrong, it's a nice feature and for an experimental piece of software nvim-treesitter works exceptionally well (and when it doesn't it's usually the parsers fault). My only suggestion was that I see it as a *plus* that lspconfig doesn't install anything, and that dedicated package managers like apt are better suited. With the Nix package manager, for example, prebuilt binaries, generated directly from the grammars, are already available. 

With 0.5 coming up, I feel like NeoVim could strongly benefit by building a cohesive API around LSP and TreeSitter that all these new Lua plugins could then standardize around. Balls in your court!

**Comment by theHamsta** [comment]
Maybe we could a wikipage to external parser installation. I can prepare something tonight.

**Comment by craftzdog** [comment]
This works:

```lua
local parser_config = require "nvim-treesitter.parsers".get_parser_configs()
parser_config.typescript.used_by = "javascript"
```

And this doesn't work:

```lua
local parser_config = require "nvim-treesitter.parsers".get_parser_configs()
parser_config.tsx.used_by = "javascript"
```

But updating `lua/nvim-treesitter/parsers.lua` itself works.
`list.tsx` already has `used_by` value by default:

```lua
list.tsx = {
  install_info = {
    url = "https://github.com/tree-sitter/tree-sitter-typescript",
    files = { "src/parser.c", "src/scanner.c" },
    location = "tree-sitter-tsx/tsx",
    generate_requires_npm = true,
  },
  used_by = { "typescript.tsx" },
  filetype = 'typescriptreact',
  maintainers = {'@steelsojka'}
}
```

and updating `used_by`  has no effect.

**Comment by JSchrtke** [comment]
I just came across this looking for something, this can probably be closed since #2471 was merged and deprecated this option anyway.
